ArduCopter/mode_auto.cpp

```diff
@@ -544,6 +546,55 @@ void ModeAuto::exit_mission()
     }
 }
 
+// detect and handle external changes to mission
+bool ModeAuto::check_for_mission_change()
+{
+    // check if mission has been updated
+    const uint32_t change_time_ms = mission.last_change_time_ms();
+    const bool update_time_changed = (change_time_ms != mis_detect.last_change_time_ms);;
+
+    // check if active command index has changed
+    const uint16_t curr_cmd_idx = mission.get_current_nav_index();
+    const bool cmd_idx_changed = (curr_cmd_idx != mis_detect.curr_cmd_index);
+
+    // no changes if neither mission update time nor active command index has changed
+    if (!update_time_changed && !cmd_idx_changed) {
+        return false;
+    }
+
+    // the mission has been updated (but maybe not changed) and/or the current command index has changed
+    // check the contents of the next three commands to ensure they have not changed
+    // and update storage so we can detect future changes
+
+    bool cmds_changed = false;  // true if upcoming command contents have changed
+
+    // retrieve cmds from mission and compare with mis_detect
+    uint8_t num_cmds = 0;
+    uint16_t cmd_idx = curr_cmd_idx;
+    AP_Mission::Mission_Command cmd[3];
+    while ((num_cmds < 3) && mission.get_next_nav_cmd(cmd_idx, cmd[num_cmds])) {
+        if ((num_cmds > mis_detect.cmd_count) || (cmd[num_cmds] != mis_detect.cmd[num_cmds])) {
+            cmds_changed = true;
+            mis_detect.cmd[num_cmds] = cmd[num_cmds];
+        }
+        cmd_idx = cmd[num_cmds].index+1;
+        num_cmds++;
+    }
+
+    // mission has changed if number of commands in mission does not match mis_detect
+    if (num_cmds != mis_detect.cmd_count) {
+        cmds_changed = true;
+    }
+
+    // update mis_detect with last change time, command index and number of commands
+    mis_detect.last_change_time_ms = change_time_ms;
+    mis_detect.curr_cmd_index = curr_cmd_idx;
+    mis_detect.cmd_count = num_cmds;
+
+    // mission has changed if command contents have changed without the current command index changing
```

Comment:
I'll change the comment to add "upcoming" so it will be, "// mission has changed if upcoming command contents have changed without the current command index changing"